### PR TITLE
Implement network explore

### PR DESCRIPTION
**Pull request by pipermerriam**
## What was wrong?

`recursive_find_nodes` gives us `O(log(n))` lookup times to find the closest node to a specific target location in the network.

When either advertising content **or** locating content, we need to both find the closest node as well as as many nodes in the immediate surrounding area. 

## How was it fixed?

Implemented `NetworkAPI.explore(target: NodeID)`

The docstring of `common_explore_network` should have a good description of the algorithm.

#### Cute Animal Picture

401 500 0_f1

### CODE REVIEW

**Comment by pipermerriam** [comment]
One thing that I'll point out about this is that it will actually enumerate all of the nodes on the network if you continue to consume the iterator which is a nice property because it gives you "guarantees" about finding the closest nodes, even if the closest nodes are really far away.

**Review by njgheorghita — APPROVED** [review]
Lookin' good! Just a couple nitpicks. It might be worth adding tests to cover some of the edge cases when prematurely exiting context (either as host / remote), even though those cases seem to be handled well.

File: ddht/v5_1/alexandria/network.py [hunk 1]
```diff
@@ -230,6 +217,11 @@ def recursive_find_nodes(
     ) -> AsyncContextManager[trio.abc.ReceiveChannel[ENRAPI]]:
         return common_recursive_find_nodes(self, target)
 
+    def explore(
+        self, target: NodeID, concurrency: int = 3,
+    ) -> AsyncContextManager[trio.abc.ReceiveChannel[ENRAPI]]:
+        return common_explore_network(self, target, concurrency=concurrency)
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply ddht/v5_1/alexandria/network.py:223]: I'm not exactly sure where your confusion lies but I think the short answer is yes.

The long answer:

```python
@asynccontextmanager
async def something() -> AsyncIterator[None]:
    do_something_before()
    yield
    do_something_after()
```

The `AsyncIterator` is to satisfy `mypy`.  The decorator transforms it roughly like this:

```python
TContextValue = TypeVar("TContextValue")

def asynccontextmanager(func: Callable[..., AsyncIterator[TContextValue]]) -> AsyncContextManager[TContextValue]:
    ...
```

So the decorator uses the `TypeVar` approach to transform the return type of the function.

As for why you're not seeing any `async` in the function definitions.

```python
@asynccontextmanager
async def my_context_fn() -> AsyncIterator[None]:
    ...

def proxy_to_my_context_fn() -> AsyncContextManager[None]:
    return my_context_fn()  # this is confusing, why don't I have to call this with `await`?????
```

The reason is that the return value of `my_context_fn()` is an instance of an object that implements the `AsyncContextManager` API:

```python
class AsyncContextManager:
    async def __aenter__(...):
    async def __aexit__(...):
```

So we don't need the `await` on the function call because python does it for us when we use it like `async with proxy_to_my_context_fn():`

So... there's no `async` needed for the definition of `NetworkAPI.explore` and no need to `await` the call to `common_explore_network`.

File: ddht/v5_1/network.py [hunk 2]
```diff
@@ -267,6 +271,274 @@ async def _monitor_done(send_channel: trio.abc.SendChannel[ENRAPI]) -> None:
...
+                            node_id.hex(),
+                            max_distance,
+                            distances[0],
+                        )
+                        break
+
+                for enr in found_enrs:
+                    try:
+                        network.enr_db.set_enr(enr)
+                    except OldSequenceNumber:
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply ddht/v5_1/network.py:403]: No, but I don't think that has been upgraded here in this repository.

File: ddht/v5_1/network.py [hunk 3]
```diff
@@ -267,6 +271,274 @@ async def _monitor_done(send_channel: trio.abc.SendChannel[ENRAPI]) -> None:
...
+        async with send_channel:
+            # First wait for the RFN to be complete.
+            await rfn_done.wait()
+
+            while True:
+                # TODO: stop-gap to ensure we don't deadlock
+                with trio.fail_after(60):
+                    async with condition:
+                        network.logger.debug(
+                            "explore: seen=%d  pending=%d  in_flight=%d  " "queried=%d",
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply ddht/v5_1/network.py:502]: This was black combining a string that was previously too long for one line that got split into two lines and then made smaller and re-combined to one line...

File: ddht/v5_1/network.py [hunk 4]
```diff
@@ -267,6 +271,274 @@ async def _monitor_done(send_channel: trio.abc.SendChannel[ENRAPI]) -> None:
...
+    rfn_done = trio.Event()
+
+    async with trio.open_nursery() as nursery:
+        nursery.start_soon(_monitor_done, send_channel, rfn_done)
+
+        nursery.start_soon(_source_nodes_from_rfn, send_channel, rfn_done)
+
+        for worker_id in range(concurrency):
+            nursery.start_soon(_worker, worker_id, send_channel)
+
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply ddht/v5_1/network.py:526]: good catch.

File: ddht/v5_1/alexandria/network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (njgheorghita) [reply ddht/v5_1/alexandria/network.py:223]: This is great, cleared everything up, thanks!
